File: IAIso-v5.0/core/iaiso-java/iaiso-consent/src/main/java/io/iaiso/consent/RevocationList.java
package io.iaiso.consent;

import java.util.Collections;
import java.util.HashSet;
import java.util.Set;
import java.util.concurrent.locks.ReentrantReadWriteLock;

/**
 * In-memory revocation list. Production deployments should back this
 * with Redis, a database, or similar.
 */
public final class RevocationList {
    private final Set<String> revoked = new HashSet<>();
    private final ReentrantReadWriteLock lock = new ReentrantReadWriteLock();

    public void revoke(String jti) {
        lock.writeLock().lock();
        try {
            revoked.add(jti);
        } finally {
            lock.writeLock().unlock();
        }
    }

    public boolean isRevoked(String jti) {
        lock.readLock().lock();
        try {
            return revoked.contains(jti);
        } finally {
            lock.readLock().unlock();
        }
    }

    public int size() {
        lock.readLock().lock();
        try {
            return revoked.size();
        } finally {
            lock.readLock().unlock();
        }
    }

    public Set<String> snapshot() {
        lock.readLock().lock();
        try {
            return Collections.unmodifiableSet(new HashSet<>(revoked));
        } finally {
            lock.readLock().unlock();
        }
    }
}

File: IAIso-v5.0/core/iaiso-swift/Sources/IAIsoConsent/RevocationList.swift
import Foundation

/// Thread-safe set of revoked JTIs.
public final class RevocationList: @unchecked Sendable {
    private let lock = NSLock()
    private var set: Set<String> = []

    public init() {}

    public func revoke(_ jti: String) {
        lock.lock(); defer { lock.unlock() }
        set.insert(jti)
    }

    public func isRevoked(_ jti: String) -> Bool {
        lock.lock(); defer { lock.unlock() }
        return set.contains(jti)
    }

    public func clear() {
        lock.lock(); defer { lock.unlock() }
        set.removeAll()
    }
}
